src/dns/query.hpp

```diff
@@ -201,13 +201,17 @@ static void dns_callback(int code, char type, int count, int ttl,
 
     context->message.answers =
             build_answers_evdns(code, type, count, ttl, addresses);
-    if (context->message.error_code != DNS_ERR_NONE) {
-        context->callback(mk::dns::dns_error(context->message.error_code),
-                context->message);
-    } else {
-        context->callback(NoError(), context->message);
+    try {
+        if (context->message.error_code != DNS_ERR_NONE) {
+            context->callback(mk::dns::dns_error(context->message.error_code),
+                    context->message);
+        } else {
+            context->callback(NoError(), context->message);
+        }
+    } catch (const std::exception& e) {
+        delete context;
+        throw;
```

Comment:
I am undecided whether we should propagate exceptions here. I think we should stop here some exceptions, namely the ones derived from `Error` that MK users would probably like to throw. Other exceptions should probably continue because typically they indicate either errors external to MK or very serious conditions.